format/midi/constants.go

```diff
@@ -368,9 +370,9 @@ var framerates = scalar.UintMapSymUint{
 	3: 30,
 }
 
-var fps = scalar.UintMapSymUint{
-	0xe8: 24,
-	0xe7: 25,
-	0xe6: 29,
-	0xe5: 30,
+var fps = scalar.SintMapSymUint{
+	-24: 24,
+	-25: 25,
+	-29: 29,
+	-30: 30,
```

Comment:
Huh what is this weird mapping? :) is stored as 2's compliment but the abs value is always used? 🤔

Reply:
Yup, welcome to the wonderful world of MIDI :-). 

However, fair point and I've remapped them to strings for clarity because 29 is not 29 - it's actually "_29.97 FPS and drop the occasional frame to keep things in sync_". **And** also discovered a bug in the _SMPTE offset_ event decoding while I was at it, so good call!

FYA: here's the relevant bit of the spec:

> If bit 15 of <division> is a one, delta-times in a file correspond to subdivisions of a second, in
> a way consistent with SMPTE and MIDI time code. Bits 14 thru 8 contain one of the four
> values -24, -25, -29, or -30, corresponding to the four standard SMPTE and MIDI time code
> formats (-29 corresponds to 30 drop frame), and represents the number of frames per second.
> These negative numbers are stored in two's complement form. The second byte (stored
> positive) is the resolution within a frame: typical values may be 4 (MIDI time code resolution),
> 8, 10, 80 (bit resolution), or 100. This system allows exact specification of time-code-based
> tracks, but also allows millisecond-based tracks by specifying 25 frames/sec and a resolution of
> 40 units per frame. If the events in a file are stored with bit resolution of thirty-frame time
> code, the division word would be E250 hex.

and:

> The new counting system counts from 0 to 29 for most seconds, but skips frame 0 and frame 1 for some 
> seconds, jumping directly from frame 29 of the previous second to frame 2 of the next second.  This counting 
> system is called “drop frame.”

Reply:
Oh 29.97 i guess from the us electric grid? same thing shows up in other digital media formats :) i guess you could use sym float if you think ppl would like to use the value in queries etc.

Reply:
It's from film to NTSC (TV) conversion - there's a whole long writeup [here](https://finale3d.com/documentation/time-code-frame-rate) if you're interested.

re. sym float. I thought about it but the rate is actually 29.97002997 **and** drop a frame so however you code it you're going to end up writing a switch statement so clarity seemed the better way to go.